project_euler/problem_191/sol1.py

```diff
@@ -0,0 +1,87 @@
+"""
+Prize Strings
+Problem 191
+
+A particular school offers cash rewards to children with good attendance and
+punctuality. If they are absent for three consecutive days or late on more
+than one occasion then they forfeit their prize.
+
+During an n-day period a trinary string is formed for each child consisting
+of L's (late), O's (on time), and A's (absent).
+
+Although there are eighty-one trinary strings for a 4-day period that can be
+formed, exactly forty-three strings would lead to a prize:
+
+OOOO OOOA OOOL OOAO OOAA OOAL OOLO OOLA OAOO OAOA
+OAOL OAAO OAAL OALO OALA OLOO OLOA OLAO OLAA AOOO
+AOOA AOOL AOAO AOAA AOAL AOLO AOLA AAOO AAOA AAOL
+AALO AALA ALOO ALOA ALAO ALAA LOOO LOOA LOAO LOAA
+LAOO LAOA LAAO
+
+How many "prize" strings exist over a 30-day period?
+
+References:
+    - The original Project Euler project page:
+      https://projecteuler.net/problem=191
+"""
+
+
+def solution(days: int = 30) -> int:
+    """Returns the number of possible prize strings for a particular number
+    of days, using a simple recursive function with caching to speed it up.
+
+    >>> solution()
+    1918080160
+    >>> solution(4)
+    43
+    """
+
+    # a cache to speed up calculation
+    cache = {}
+
+    # we will be using a simple recursive function
+    def calculate(days: int, absent: int, late: int) -> int:
```

Comment:
Please separate out this function and include `doctest` and `docstring` for it.

Reply:
I wanted to keep the helper function "inside" the solution function, since it is an implementation detail and not part of the (public) interface. Is this not usually done in the python world?

Reply:
Oh no, you don't actually need to hide the implementation details here but in the real world if you want to achieve that then you use underscores. You can google it up as it's quite a deep dive. Also, function inside a function is mostly used just for a wrapper function (you can google that up as well).

Reply:
but this basically *is* just a wrapper function, as the "solution()" function is now pretty much  just a 1-liner